### DISCUSSION TITLE
How to get a list of all positions in a file?

### DISCUSSION THREAD

**Discussion by weilbith** [nvim-neotest/neotest#183] [Q&A]
Hey ✌

I read a while through the code, but I fail to figure out a plain method to call that gives me a list of all positions (of test functions) in a file. Without knowing the adpater or so.
Maybe just the row. Better also with column or even range. If it would also include the result it can't get better. 😅

Any hints?

**Comment by rcarriga** [comment]
Right now, none of the consumers provide this information so you'd have to write your own. It's slightly more involved than a direct API call. The main thing is to use the async module to wrap calls because they are mostly async on the client.

To use a custom consumer you can provide them in the setup:
```lua
    consumers = {
      my_consumer = function (client)
      end
    },
```
I've just pushed some doc updates and extra type hints for a lot of neotest so hopefully there is enough info in the docs on how to use the client along with existing consumers as examples. The `run` consumer is probably a good place to start as it's quite simple
https://github.com/nvim-neotest/neotest/blob/de2f68fb13a05c361e0c096e167f7f79712f3a5c/lua/neotest/consumers/run.lua#L19-L39

I'd heavily recommend using something like https://github.com/folke/neodev.nvim to enable type checking of neotest, it will make life a lot easier (and be sure to pull the latest commits) :sweat_smile: 

To get the tests in a file, you're looking for `:h neotest.Client.get_position()`

**Reply by weilbith** [reply]
I was actually hoping to not use a custom provider. 😅

Yes, this position function seems good, but I failed to understand how to use it with the position identifier etc. 🙈

**Reply by rcarriga** [reply]
Right now as I said there's no other way unfortunately but there will hopefully be a consumer to help with this soon. This functionality would fit in nicely by extending the consumer API I proposed here https://github.com/nvim-neotest/neotest/pull/134#issuecomment-1315827513 and would allow more simple interaction without needing to do async work.

The position IDs for directories and files are just their absolute paths so for the current buffer you can just use `vim.fn.expand("%:p")` as the position ID in `client:get_position(position_id)`
